java/pattern/constraint/ValueConstraint.java

```diff
@@ -0,0 +1,663 @@
+
+                private final Operation.OP op;
+                private final Expression a;
+                private final Expression b;
+                private final int hash;
+
+                public Operation(Operation.OP op, Expression a, Expression b) {
+                    this.op = op;
+                    this.a = a;
+                    this.b = b;
+                    this.hash = Objects.hash(op, a, b);
+                }
+
+                public Operation.OP operator() {
+                    return op;
+                }
+
+                public Pair<Expression, Expression> operands() {
+                    return new Pair<>(a, b);
+                }
+
+                @Override
+                public boolean isOperation() {
+                    return true;
+                }
+
+                @Override
+                public Operation asOperation() {
+                    return this;
+                }
+
+                @Override
+                protected void collectVariables(Set<BoundVariable> collector) {
+                    a.collectVariables(collector);
+                    b.collectVariables(collector);
+                }
+
+                @Override
+                public String toString() {
+                    return a.toString() + " " + op.symbol + " " + b.toString();
+                }
+
+                @Override
+                public boolean equals(Object o) {
+                    if (this == o) return true;
+                    if (o == null || getClass() != o.getClass()) return false;
+                    Operation that = (Operation) o;
+                    return (this.op.equals(that.op) && this.a.equals(that.a) && this.b.equals(that.b));
+                }
+
+                @Override
+                public int hashCode() {
+                    return hash;
+                }
+            }
+
+            public static class Function extends Expression {
+                private final String symbol;
+                private final List<Expression> argList;
+                private final int hash;
+
+                public Function(String symbol, List<Expression> argList) {
+                    this.symbol = symbol;
+                    this.argList = argList;
+                    this.hash = Objects.hash(this.symbol, this.argList);
+                }
+
+                public String symbol() {
+                    return symbol;
+                }
+
+                public List<Expression> arguments() {
+                    return argList;
+                }
+
+                @Override
+                public boolean isFunction() {
+                    return true;
+                }
+
+                @Override
+                public Function asFunction() {
+                    return this;
+                }
+
+                @Override
+                protected void collectVariables(Set<BoundVariable> collector) {
+                    argList.forEach(arg -> arg.collectVariables(collector));
```

Comment:
i think you can just call it `args` in here, since it's gotta be a list and we don't tend to put the data structure type in the variable name

Reply:
It's recursive though. And collector makes it a bit more explicit that it's an 'out' variable.
Calling it args sounds like it's an input argument?

Reply:
Oh you meant the local data structure. Sorry I misunderstood. I'll make it args.